querier/querier.go

```diff
@@ -206,6 +201,97 @@ func (qm MergeQuerier) Close() error {
 	return nil
 }
 
+// RemoteReadHandler handles Prometheus remote read requests.
+func (qm MergeQuerier) RemoteReadHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	var req cortex.ReadRequest
+	if err := distributor.ParseProtoRequest(ctx, w, r, &req, true); err != nil {
+		log.Errorf(err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// Fetch samples for all queries in parallel.
+	matrices := make(chan model.Matrix)
+	errors := make(chan error)
+	for _, q := range req.Queries {
+		go func(q *cortex.QueryRequest) {
+			from, to, matchers, err := util.FromQueryRequest(q)
+			if err != nil {
+				errors <- err
+				return
+			}
+
+			matrix, err := qm.Query(ctx, from, to, matchers...)
+			if err != nil {
+				errors <- err
+			} else {
+				matrices <- matrix
+			}
+		}(q)
+	}
+
+	matrix, err := mergeMatrices(matrices, errors, len(req.Queries))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		log.Errorf("error executing remote read request: %v", err)
+		return
+	}
+
+	resp := util.ToQueryResponse(matrix)
+
+	data, err := proto.Marshal(resp)
```

Comment:
Can you put the marshalling logic in a separate function `SerialiseProtoResponse` or somthing please?